presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -354,54 +360,41 @@ private void invokeNoMoreSplitsIfNecessary()
                 throw new PrestoException(NOT_SUPPORTED, "Bucketed table in SymlinkTextInputFormat is not yet supported");
             }
-
-            // TODO: This should use an iterator like the HiveFileIterator
-            ListenableFuture<?> lastResult = COMPLETED_FUTURE;
-            List<Path> targetPaths = hdfsEnvironment.doAs(
+            FileStatus[] symlinks = hdfsEnvironment.doAs(
                     hdfsContext.getIdentity().getUser(),
-                    () -> getTargetPathsFromSymlink(fs, path));
-            for (Path targetPath : targetPaths) {
-                // The input should be in TextInputFormat.
-                TextInputFormat targetInputFormat = new TextInputFormat();
-                // the splits must be generated using the file system for the target path
-                // get the configuration for the target path -- it may be a different hdfs instance
-                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, targetPath);
-                JobConf targetJob = toJobConf(targetFilesystem.getConf());
-                targetJob.setInputFormat(TextInputFormat.class);
-                Optional<Principal> principal = hdfsContext.getIdentity().getPrincipal();
-                if (principal.isPresent()) {
-                    targetJob.set(MRConfig.FRAMEWORK_NAME, MRConfig.CLASSIC_FRAMEWORK_NAME);
-                    targetJob.set(MRConfig.MASTER_USER_NAME, principal.get().getName());
-                }
-                targetInputFormat.configure(targetJob);
-                FileInputFormat.setInputPaths(targetJob, targetPath);
-                InputSplit[] targetSplits = hdfsEnvironment.doAs(
-                        hdfsContext.getIdentity().getUser(),
-                        () -> targetInputFormat.getSplits(targetJob, 0));
+                    () -> fs.listStatus(path, HIDDEN_FILES_PATH_FILTER));
 
+            InputFormat<?, ?> fileInputFormat = mapSymlinkSerdeToInputFormat(toJobConf(configuration), getDeserializerClassName(schema));
+
+            for (FileStatus symlink : symlinks) {
                 InternalHiveSplitFactory splitFactory = new InternalHiveSplitFactory(
-                        targetFilesystem,
+                        fs,
                         partitionName,
-                        inputFormat,
+                        fileInputFormat,
                         schema,
                         partitionKeys,
                         effectivePredicate,
                         partitionMatchSupplier,
                         partition.getTableToPartitionMapping(),
-                        Optional.empty(),
+                        Optional.empty(), //no bucketing support
                         getMaxInitialSplitSize(session),
                         isForceLocalScheduling(session),
                         s3SelectPushdownEnabled);
-                lastResult = addSplitsToSource(targetSplits, splitFactory);
-                if (stopped) {
```

Comment:
this was breaking potentially expensive listing when query is cancelled. Could we preserve it (e.g after listing files for each parent)?